dspace-api/src/main/java/org/dspace/uclouvain/core/mail/MailMetadataParserServiceImpl.java

```diff
@@ -0,0 +1,191 @@
+     * @param metadataFields The fields to extract from the item.
+     * @param lang The language used to generate the labels.
+     * @return A map containing all the labels and values for the given metadata fields.
+     * The map is of type {@link LinkedHashMap} which allows to keep the order of the metadataFields list.
+     */
+    public LinkedHashMap<String, String> parseMetadata(
+        Context context, Item item, List<String> metadataFields, String lang
+    ) {
+        return metadataFields.stream()
+            .map(field -> parseMetadata(context, item, field, lang))
+            .filter(pair -> pair != null)
+            // Reduce here to convert from List<Pair<String, String>> to LinkedHashMap<String, String>
+            .reduce(
+                new LinkedHashMap<>(),
+                (acc, pair) -> {
+                    acc.put(pair.getLeft(), pair.getRight());
+                    return acc;
+                },
+                (map1, map2) -> {
+                    map1.putAll(map2);
+                    return map1;
+                }
+            );
+    }
+
+    /**
+     * Retrieve the correct parser for the given metadata field name.
+     * @param metadataField The metadata field name of form '<schema>.<element>.<qualifier>'.
+     * @return The parser configured to handle the metadata field or null if not found.
+     */
+    private MailMetadataParser getParserForMetadataField(String metadataField) {
+        return parsers.stream()
+            .filter(parser -> parser.getMetadataField().equals(metadataField))
+            .findFirst()
+            .orElse(null);
+    }
+
+    /**
+     * Get the values for the given item, metadata field and parser.
+     * 
+     * @param context The current DSpace context.
+     * @param item The item to extract initial values from.
+     * @param parser The parser used to parse the values.
+     * @param metadataField The metadata field to extract values from the item.
+     * @return Returns a parsed value. Can return an empty string if no values were found for the given metadata field.
+     */
+    private String getValue(Context context, Item item, MailMetadataParser parser, String metadataField) {
+        List<String> values = itemService.getMetadataByMetadataString(item, metadataField).stream()
+            .map(MetadataValue::getValue).collect(Collectors.toList());
+        if (values.isEmpty()) {
+            // Return an empty string if no values were found.
+            return "";
+        }
+
+        // If multipleValues flag is false, just get the first metadata value.
+        if (!parser.getMultipleValues()) {
+            values = Arrays.asList(values.get(0));
+        }
+        // If we have a configured vocabulary name, we need to try to convert the values.
+        if (parser.getVocabularyName() != null) {
+            values = convertVocabularyValues(context, values, parser);
+        }
+        return String.join(parser.getSeparator(), values);
```

Comment:
There is a bug here. If parser is linked to a vocabulary but for a singleValue, the value return will not be converted.
For me all this logic should be moves into the `MailMetadataParser` class with implementation of a `getValue(context, values)` function